woocommerce/woocommerce
# Support for an external REST API package hosted on Packagist

This PR implements the REST API as an external package hosted on Packagist (https://packagist.org/packages/woocommerce/woocommerce-rest-api) and developed on [Github here](https://github.com/woocommerce/woocommerce-rest-api).

**Why?**
This allows dev teams to collaborate on a single REST API repository and run development versions of the REST API either as a feature plugin, or as a plugin dependency, independent of the version in WooCommerce core.

**About the package**
The package ([developed in the woocommerce-rest-api repository](https://github.com/woocommerce/woocommerce-rest-api)) runs standalone as a feature plugin, but can also be included in other plugins as a composer dependency. 

This PR includes functionality to register versions of the package and allows core to **only load the latest version of the package**. If no other versions are registered, core will load the core (presumably stable) version.

Versions are registered by providing a callback function and version number. Including the main file (`woocommerce-rest-api.php`) or doing this manually are both viable options. The code looks like this:

```
/**
 * Get API feature plugin version and callback function.
 */
$version        = include __DIR__ . '/version.php';
$init_callback = include __DIR__ . '/init.php';

/**
 * This callback registers this version of the API with WooCommerce.
 */
$register_callback = function() use ( $version, $init_callback ) {
	if ( ! is_callable( array( wc()->api, 'register' ) ) ) {
		return;
	}
	wc()->api->register( $version, $init_callback );
};

add_action( 'woocommerce_loaded', $register_callback );
```

The `version` and `init` files return anonymous functions; version gives you the hardcoded version number of the package, and init provides a callback to load the package and its autoloader. Once a package is loaded, thats it; the autoloader takes care of including the correct versions of the REST API. Other registered versions are never included/called.

**Build step**
The REST API is pulled in using composer. If the version in core is missing, a notice will shown in admin asking that you run the build step. That involves going to the `woocommerce` plugin directory and running `npm run build` or `composer install`. This adds the package to the `includes/rest-api` directory.

_Important:_ `includes/rest-api`  is excluded from source control. This needs to be build prior to deployments to WordPress.org so our production version of WooCommerce has the REST API included.

**Testing**
- Checkout branch; see a notice in admin asking you to perform the build step
- Follow instructions; run the build step. See the `rest-api` directory appear in `woocommerce/includes/` - notice should be removed.
- Check System status - the REST API version is indicated there. For now it's 1.1 but this is just for testing and will change.
- Try accessing some REST endpoints - spot check - there should be no issues. Same for legacy API which was simply moved (the legacy API is not part of the new package).
- Download the package directory and install as a plugin. [Repo is here](https://github.com/woocommerce/woocommerce-rest-api)).
- Ensure you can activate the plugin and the API continues to work. The version will differ in time, right now both versions are 1.1 and there are no tagged versions.

Requesting feedback.
